invenio_rdm_records/services/communities/components.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Record communities service components."""
+
+from invenio_communities.communities.records.systemfields.access import (
+    CommunityAccess,
+    VisibilityEnum,
+)
+from invenio_communities.communities.services.components import (
+    CommunityAccessComponent as BaseAccessComponent,
+)
+from invenio_communities.communities.services.components import (
+    CustomFieldsComponent,
+    FeaturedCommunityComponent,
+    OAISetComponent,
+    OwnershipComponent,
+    PIDComponent,
+)
+from invenio_records_resources.services.records.components import (
+    MetadataComponent,
+    RelationsComponent,
+)
+from invenio_search.engine import dsl
+
+from ...proxies import current_community_records_service
+from ..errors import InvalidCommunityVisibility
+
+
+class CommunityAccessComponent(BaseAccessComponent):
+    """Community access component."""
+
+    def _check_visibility(self, identity, record):
+        """Checks if the visibility change is allowed."""
+        if VisibilityEnum(record.access.visibility) == VisibilityEnum.RESTRICTED:
+            assert CommunityAccess.validate_visibility_level(VisibilityEnum.PUBLIC)
+
+            total = current_community_records_service.search(
+                identity,
+                community_id=record.id,
+                extra_filter=dsl.Q(
+                    "term", **{"access.record": str(VisibilityEnum.PUBLIC)}
+                ),
+            ).total
+            if total > 0:
+                raise InvalidCommunityVisibility(
+                    reason="Cannot restrict a community with public records."
```

Comment:
I can add this in my next PR ;)
```suggestion
                    reason=_("Cannot restrict a community with public records.")
```